# Laporan Praktikum 10 Praktik Sistem Kendali Cerdas

# Liana Dewi Ramadanty

# 19507334035

In [2]:
import anfis
import membership.mfDerivs
import membership.membershipfunction
import numpy

ModuleNotFoundError: No module named 'membership'

In [4]:
ts = numpy.loadtxt("trainingSet.txt", usecols=[1,2,3])#numpy.loadtxt('c:\\Python_fiddling\\myProject\\MF\\trainingSet.txt',usecols=[1,2,3])
ts

array([[-1.00000000e+01, -1.00000000e+01,  2.95958969e-03],
       [-1.00000000e+01, -8.00000000e+00, -6.72789715e-03],
       [-1.00000000e+01, -6.00000000e+00,  2.53346550e-03],
       [-1.00000000e+01, -4.00000000e+00,  1.02929134e-02],
       [-1.00000000e+01, -2.00000000e+00, -2.47338498e-02],
       [-1.00000000e+01,  0.00000000e+00, -5.44021111e-02],
       [-1.00000000e+01,  2.00000000e+00, -2.47338498e-02],
       [-1.00000000e+01,  4.00000000e+00,  1.02929134e-02],
       [-1.00000000e+01,  6.00000000e+00,  2.53346550e-03],
       [-1.00000000e+01,  8.00000000e+00, -6.72789715e-03],
       [-1.00000000e+01,  1.00000000e+01,  2.95958969e-03],
       [-8.00000000e+00, -1.00000000e+01, -6.72789715e-03],
       [-8.00000000e+00, -8.00000000e+00,  1.52942147e-02],
       [-8.00000000e+00, -6.00000000e+00, -5.75920890e-03],
       [-8.00000000e+00, -4.00000000e+00, -2.33983997e-02],
       [-8.00000000e+00, -2.00000000e+00,  5.62263067e-02],
       [-8.00000000e+00,  0.00000000e+00

Ts merupakan data yang ada di trainingSet.Txt

In [5]:
X = ts[:,0:2]
X

array([[-10., -10.],
       [-10.,  -8.],
       [-10.,  -6.],
       [-10.,  -4.],
       [-10.,  -2.],
       [-10.,   0.],
       [-10.,   2.],
       [-10.,   4.],
       [-10.,   6.],
       [-10.,   8.],
       [-10.,  10.],
       [ -8., -10.],
       [ -8.,  -8.],
       [ -8.,  -6.],
       [ -8.,  -4.],
       [ -8.,  -2.],
       [ -8.,   0.],
       [ -8.,   2.],
       [ -8.,   4.],
       [ -8.,   6.],
       [ -8.,   8.],
       [ -8.,  10.],
       [ -6., -10.],
       [ -6.,  -8.],
       [ -6.,  -6.],
       [ -6.,  -4.],
       [ -6.,  -2.],
       [ -6.,   0.],
       [ -6.,   2.],
       [ -6.,   4.],
       [ -6.,   6.],
       [ -6.,   8.],
       [ -6.,  10.],
       [ -4., -10.],
       [ -4.,  -8.],
       [ -4.,  -6.],
       [ -4.,  -4.],
       [ -4.,  -2.],
       [ -4.,   0.],
       [ -4.,   2.],
       [ -4.,   4.],
       [ -4.,   6.],
       [ -4.,   8.],
       [ -4.,  10.],
       [ -2., -10.],
       [ -2.,  -8.],
       [ -2.,  -6.],
       [ -2.,

Data X diatas merupakan data input yang diambil dari data Ts

In [6]:
Y = ts[:,2]
Y

array([ 0.00295959, -0.0067279 ,  0.00253347,  0.01029291, -0.02473385,
       -0.05440211, -0.02473385,  0.01029291,  0.00253347, -0.0067279 ,
        0.00295959, -0.0067279 ,  0.01529421, -0.00575921, -0.0233984 ,
        0.05622631,  0.12366978,  0.05622631, -0.0233984 , -0.00575921,
        0.01529421, -0.0067279 ,  0.00253347, -0.00575921,  0.0021687 ,
        0.00881093, -0.02117265, -0.04656925, -0.02117265,  0.00881093,
        0.0021687 , -0.00575921,  0.00253347,  0.01029291, -0.0233984 ,
        0.00881093,  0.03579688, -0.08601982, -0.18920062, -0.08601982,
        0.03579688,  0.00881093, -0.0233984 ,  0.01029291, -0.02473385,
        0.05622631, -0.02117265, -0.08601982,  0.20670545,  0.45464871,
        0.20670545, -0.08601982, -0.02117265,  0.05622631, -0.02473385,
       -0.05440211,  0.12366978, -0.04656925, -0.18920062,  0.45464871,
        1.        ,  0.45464871, -0.18920062, -0.04656925,  0.12366978,
       -0.05440211, -0.02473385,  0.05622631, -0.02117265, -0.08

Y adalah data untuk menampilkan data output pada data Ts

In [7]:
mf = [[['gaussmf',{'mean':0.,'sigma':1.}],
       ['gaussmf',{'mean':-1.,'sigma':2.}],
       ['gaussmf',{'mean':-4.,'sigma':10.}],
       ['gaussmf',{'mean':-7.,'sigma':7.}]],
        [['gaussmf',{'mean':1.,'sigma':2.}],
        ['gaussmf',{'mean':2.,'sigma':3.}],
        ['gaussmf',{'mean':-2.,'sigma':10.}],
        ['gaussmf',{'mean':-10.5,'sigma':5.}]]]
mf

[[['gaussmf', {'mean': 0.0, 'sigma': 1.0}],
  ['gaussmf', {'mean': -1.0, 'sigma': 2.0}],
  ['gaussmf', {'mean': -4.0, 'sigma': 10.0}],
  ['gaussmf', {'mean': -7.0, 'sigma': 7.0}]],
 [['gaussmf', {'mean': 1.0, 'sigma': 2.0}],
  ['gaussmf', {'mean': 2.0, 'sigma': 3.0}],
  ['gaussmf', {'mean': -2.0, 'sigma': 10.0}],
  ['gaussmf', {'mean': -10.5, 'sigma': 5.0}]]]

Gaussmf memiliki nilai mean dan nilai sigma. Nilai mean adalah nilai center, nilai sigma adalah nilai standar deviasi. Kemudian kita simpan dalam variabel mf.

In [8]:
mfc = membership.membershipfunction.MemFuncs(mf)
mfc

Kode diatas adalah perintah untuk mencetak membershipfunction layer menggunakan perintah membershipfunction.MemFuncs(mf)

In [10]:
anf = anfis.ANFIS(X, Y, mfc)
anf

Untuk menjalankan anfis, kita hanya tinggal mamasukkan nilai x sebagai input, y sebagai output, dan mfc yang merupakan bentuk dari membershipfunction yang digunakan diatas

In [11]:
anf.trainHybridJangOffLine(epochs=20)

current error: 0.3219713313023792
current error: 0.3206496967785757
current error: 0.3193598359657091
current error: 0.3180964819660266
current error: 0.31673125060829765
current error: 0.3152546970543038
current error: 0.31365648102556865
current error: 0.31192543517961857
current error: 0.31004976930416217
current error: 0.30801749864543804
current error: 0.3058172226718672
current error: 0.30343942051130285
current error: 0.300878450478208
current error: 0.29813538447370813
current error: 0.29522149867013137
current error: 0.29216105234630596
current error: 0.28898684515632633
current error: 0.2856912569145435
current error: 0.28184647973417154


array([[-1.19786767e-02],
       [ 5.19007695e-03],
       [ 1.50858291e-02],
       [ 5.66106857e-03],
       [-1.66125571e-02],
       [ 2.58351102e-03],
       [ 3.35328470e-02],
       [ 2.84855950e-02],
       [ 1.63883100e-03],
       [-1.66157597e-03],
       [ 5.15880107e-03],
       [-1.03492980e-02],
       [ 6.03637357e-03],
       [ 1.49551447e-02],
       [ 5.02809128e-03],
       [-2.20156604e-02],
       [-1.51514181e-02],
       [ 1.22868327e-02],
       [ 1.84446227e-02],
       [ 4.90866622e-03],
       [ 5.85381400e-04],
       [ 1.93971607e-03],
       [-7.45887383e-03],
       [ 7.98275085e-03],
       [ 1.35220824e-02],
       [ 2.54597152e-04],
       [-3.13692042e-02],
       [-4.34094757e-02],
       [-3.16114076e-02],
       [-1.31716722e-02],
       [ 2.32403948e-03],
       [ 3.63461095e-03],
       [-1.97031364e-03],
       [-2.29345985e-03],
       [ 1.24549606e-02],
       [ 4.90763465e-03],
       [-2.03049396e-02],
       [-2.14149558e-02],
       [-2.0

Perintah diatas berfungsi untuk mentraining dengan parameter (epochs=20) dimana ada 20 kali iterasi untuk menghasilkan nilai error.

In [1]:
import anfis
import membership.mfDerivs
import membership.membershipfunction
import numpy

ts = numpy.loadtxt("trainingSet.txt", usecols=[1,2,3])#numpy.loadtxt('c:\\Python_fiddling\\myProject\\MF\\trainingSet.txt',usecols=[1,2,3])
X = ts[:,0:2]
Y = ts[:,2]

mf = [[['gaussmf',{'mean':0.,'sigma':1.}],
       ['gaussmf',{'mean':-1.,'sigma':2.}],
       ['gaussmf',{'mean':-4.,'sigma':10.}],
       ['gaussmf',{'mean':-7.,'sigma':7.}]],
    [['gaussmf',{'mean':1.,'sigma':2.}],
     ['gaussmf',{'mean':2.,'sigma':3.}],
     ['gaussmf',{'mean':-2.,'sigma':10.}],
     ['gaussmf',{'mean':-10.5,'sigma':5.}]]]


mfc = membership.membershipfunction.MemFuncs(mf)
anf = anfis.ANFIS(X, Y, mfc)
anf.trainHybridJangOffLine(epochs=20)

#print(round(anf.consequents[-1][0],6))
#print(round(anf.consequents[-2][0],6))
#print(round(anf.fittedValues[9][0],6))
#if round(anf.consequents[-1][0],6) == -5.275538 and round(anf.consequents[-2][0],6) == -1.990703 and round(anf.fittedValues[9][0],6) == 0.002249:
#	print('test is good')

print("Plotting errors")
anf.plotErrors()
print("Plotting results")
anf.plotResults()

ModuleNotFoundError: No module named 'membership'

Plotting error diatas terlihat pada saat awal training, nilai error sekitar 0.3219713313023792, tetapi setelah proses 20 kali, nilai error menjadi 0.28184647973417154. Hal tersebut menunjukkan bahwa terdapat perbaikan dari sisi errornya karena nilai error semakin mengecil. Jika proses diperpanjang, misal kita ubah lebih dari 20 kali proses, kemungkinan error akan semakin mengecil bahkan akan mendekati 0.

Plotting Result adalah gambar perbandingan antara sinyal keluaran (original/sinyal biru) dengan sinyal hasil trained (sinyal berwarna merah) / output yang diprediksi yang berdasarkan hasil training. Dilihat dari grafik yang dihasilkan memang belum sempurna, tetapi dalam grafik tersebut, bisa dilihat bahwa nilainya tidak terlalu jauh dari hasil original. Sinyal trained bisa mengikuti sinyal original walaupun ada beberapa titik-titik yang mungkin sedikit terlambat/ nilainya tidak sesuai dengan hasil originalnya. Hal tersebut dapat dilakukan perbaikan agar nilai hasil trained-nya mendekati nilai original. 